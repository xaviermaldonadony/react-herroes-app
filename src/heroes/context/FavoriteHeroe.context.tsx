import type { Hero } from '../types/hero.interface';
import {
  createContext,
  useEffect,
  useState,
  type PropsWithChildren,
} from 'react';

interface FavoriteHeroContext {
  // state
  favorites: Hero[];
  favoriteCount: number;

  //   methods
  isFavorite: (hero: Hero) => boolean;
  toggleFavorites: (hero: Hero) => void;
}
// eslint-disable-next-line react-refresh/only-export-components
export const FavoriteHeroContext = createContext({} as FavoriteHeroContext);

const getFavoritesFromLocalStorage = (): Hero[] => {
  const favorites = localStorage.getItem('favorites');
  return favorites ? JSON.parse(favorites) : [];
};

export const FavoriteHeroeProvider = ({ children }: PropsWithChildren) => {
  const [favorites, setFavorites] = useState<Hero[]>(
    getFavoritesFromLocalStorage()
  );

  const toggleFavorites = (hero: Hero) => {
    const heroExist = favorites.find((h) => h.id === hero.id);

    if (heroExist) {
      setFavorites(favorites.filter((h) => h.id !== hero.id));
      return;
    }

    setFavorites([...favorites, hero]);
  };

  const isFavorite = (hero: Hero) => favorites.some((h) => h.id === hero.id);

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(favorites));
  }, [favorites]);

  return (
    <FavoriteHeroContext
      value={{
        favoriteCount: favorites.length,
        favorites,

        // methods
        isFavorite,
        toggleFavorites,
      }}
    >
      {children}
    </FavoriteHeroContext>
  );
};
