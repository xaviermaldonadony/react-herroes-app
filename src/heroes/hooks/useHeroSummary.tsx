import { useQuery } from '@tanstack/react-query';
import { getSummaryAcion } from '../actions/get-summary-actions';

export const useHeroSummary = () => {
  return useQuery({
    queryKey: ['summary-information'],
    queryFn: getSummaryAcion,
    staleTime: 1000 * 60 * 5, // 5 mins
  });
};
