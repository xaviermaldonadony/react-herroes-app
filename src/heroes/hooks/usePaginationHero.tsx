import { useQuery } from '@tanstack/react-query';
import { getHeroesByPageAction } from '../actions/get-heroes-by-page.action';

export const usePaginationHero = (
  page: number,
  limit: number,
  category = 'all'
) => {
  return useQuery({
    queryKey: ['heroes', { page, limit, category }],
    queryFn: () => getHeroesByPageAction(+page, +limit, category),
    staleTime: 1000 * 60 * 5, // 5 minutes
  });
};
