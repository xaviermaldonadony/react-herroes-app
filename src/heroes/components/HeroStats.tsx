import { use } from 'react';
import { Badge } from '@/components/ui/badge';

import { Users, Heart, Zap, Trophy } from 'lucide-react';
import { HeroStatCard } from './HeroStatCard';
import { useHeroSummary } from '../hooks/useHeroSummary';
import { FavoriteHeroContext } from '../context/FavoriteHeroe.context';

export const HeroStats = () => {
  const { data: summary } = useHeroSummary();
  const { favoriteCount } = use(FavoriteHeroContext);

  if (!summary) {
    return <div>Loading...</div>;
  }

  const percetnageFavorite = (
    (favoriteCount / summary.totalHeroes) *
    100
  ).toFixed(2);

  return (
    <div className='grid grid-cols-2 md:grid-cols-4 gap-4 mb-8'>
      <HeroStatCard
        title={'Total Characters'}
        icon={<Users className='h-4 w-4 text-muted-foreground' />}
      >
        <div className='text-2xl font-bold'>{summary?.totalHeroes}</div>
        <div className='flex gap-1 mt-2'>
          <Badge variant='secondary' className='text-xs'>
            {summary?.heroCount} Heroes
          </Badge>
          <Badge variant='destructive' className='text-xs'>
            {summary?.villainCount} Villains
          </Badge>
        </div>
      </HeroStatCard>

      <HeroStatCard
        title={'Favorite'}
        icon={<Heart className='h-4 w-4 text-muted-foreground' />}
      >
        {/* TODO */}
        <div className='text-2xl font-bold text-red-600'>{favoriteCount}</div>
        <p className='text-xs text-muted-foreground'>{percetnageFavorite}%</p>
      </HeroStatCard>

      <HeroStatCard
        title='Strongest'
        icon={<Zap className='h-4 w-4 text-muted-foreground' />}
      >
        <div className='text-lg font-bold'>{summary?.strongestHero.alias}</div>
        <p className='text-xs text-muted-foreground'>
          Strength:{summary?.strongestHero.strength}/10
        </p>
      </HeroStatCard>

      <HeroStatCard
        title={'Smartest'}
        icon={<Trophy className='h-4 w-4 text-muted-foreground' />}
      >
        <div className='text-lg font-bold'>{summary?.smartestHero.alias}</div>
        <p className='text-xs text-muted-foreground'>
          Intelligence: {summary?.smartestHero.intelligence}/10
        </p>
      </HeroStatCard>
    </div>
  );
};
