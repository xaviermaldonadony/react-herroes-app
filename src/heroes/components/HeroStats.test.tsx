import { describe, expect, test, vi } from 'vitest';
import { HeroStats } from './HeroStats';
import { render, screen } from '@testing-library/react';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { useHeroSummary } from '../hooks/useHeroSummary';
import type { SummaryInformationResponse } from '../types/summary-information.response';
import { FavoriteHeroeProvider } from '../context/FavoriteHeroe.context';

vi.mock('../hooks/useHeroSummary');
const mockUseHeroSummary = vi.mocked(useHeroSummary);

const mockHero = {
  id: '1',
  name: 'Clark Kent',
  slug: 'clark-kent',
  alias: 'Superman',
  powers: [
    'Super strength',
    'Flight',
    'Heat vision',
    'X-ray vision',
    'Invulnerability',
    'Super speed',
  ],
  description:
    'The Last Son of Krypton, protector of Earth and a symbol of hope for all humanity.',
  strength: 10,
  intelligence: 8,
  speed: 9,
  durability: 10,
  team: 'Justice League',
  image: '1.jpeg',
  firstAppearance: '1938',
  status: 'Active',
  category: 'Hero',
  universe: 'DC',
};

const mockSummaryData: SummaryInformationResponse = {
  totalHeroes: 25,
  strongestHero: {
    id: '1',
    name: 'Clark Kent',
    slug: 'clark-kent',
    alias: 'Superman',
    powers: [
      'Super strength',
      'Flight',
      'Heat vision',
      'X-ray vision',
      'Invulnerability',
      'Super speed',
    ],
    description:
      'The Last Son of Krypton, protector of Earth and a symbol of hope for all humanity.',
    strength: 10,
    intelligence: 8,
    speed: 9,
    durability: 10,
    team: 'Justice League',
    image: '1.jpeg',
    firstAppearance: '1938',
    status: 'Active',
    category: 'Hero',
    universe: 'DC',
  },
  smartestHero: {
    id: '2',
    name: 'Bruce Wayne',
    slug: 'bruce-wayne',
    alias: 'Batman',
    powers: [
      'Martial arts',
      'Detective skills',
      'Advanced technology',
      'Stealth',
      'Tactical genius',
    ],
    description:
      'The Dark Knight of Gotham City, who uses fear as a weapon against crime and corruption.',
    strength: 6,
    intelligence: 10,
    speed: 6,
    durability: 7,
    team: 'Justice League',
    image: '2.jpeg',
    firstAppearance: '1939',
    status: 'Active',
    category: 'Hero',
    universe: 'DC',
  },
  heroCount: 18,
  villainCount: 7,
};
const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});

const renderHeroStats = (mockData?: Partial<SummaryInformationResponse>) => {
  mockUseHeroSummary.mockReturnValue({
    data: mockData ?? undefined,
  } as unknown as ReturnType<typeof useHeroSummary>);

  return render(
    <QueryClientProvider client={queryClient}>
      <FavoriteHeroeProvider>
        <HeroStats />
      </FavoriteHeroeProvider>
    </QueryClientProvider>
  );
};

describe('HeroStats', () => {
  test('should render component with default values', () => {
    const { container } = renderHeroStats();

    expect(screen.getByText('Loading...')).toBeDefined();
    expect(container).toMatchSnapshot();
  });

  test('should render HeroStats with mock data', () => {
    const { container } = renderHeroStats(mockSummaryData);

    expect(container).toMatchSnapshot();
    expect(screen.getByText('Total Characters')).toBeDefined();
    expect(screen.getByText('Favorite')).toBeDefined();
    expect(screen.getByText('Strongest')).toBeDefined();
  });

  test('should change the percentage of favorites when hero is added to favorites', () => {
    localStorage.setItem('favorites', JSON.stringify([mockHero]));
    renderHeroStats(mockSummaryData);

    const favoritePercentageElement = screen.getByTestId('favorite-percentage');
    const favoriteCountElement = screen.getByTestId('favorite-count');
    expect(favoritePercentageElement.innerHTML).toContain('4.00%');
    expect(favoriteCountElement.innerHTML).toContain('1');
  });
});
