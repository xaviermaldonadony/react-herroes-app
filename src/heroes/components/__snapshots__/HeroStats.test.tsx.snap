// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HeroStats > should render HeroStats with mock data 1`] = `
<div>
  <div
    class="grid grid-cols-2 md:grid-cols-4 gap-4 mb-8"
  >
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Total Characters
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-users h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
          />
          <path
            d="M16 3.128a4 4 0 0 1 0 7.744"
          />
          <path
            d="M22 21v-2a4 4 0 0 0-3-3.87"
          />
          <circle
            cx="9"
            cy="7"
            r="4"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold"
        >
          25
        </div>
        <div
          class="flex gap-1 mt-2"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
            18
             Heroes
          </span>
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-destructive text-white [a&]:hover:bg-destructive/90 focus-visible:ring-destructive/20 dark:focus-visible:ring-destructive/40 dark:bg-destructive/60 text-xs"
            data-slot="badge"
          >
            7
             Villains
          </span>
        </div>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Favorite
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-heart h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M2 9.5a5.5 5.5 0 0 1 9.591-3.676.56.56 0 0 0 .818 0A5.49 5.49 0 0 1 22 9.5c0 2.29-1.5 4-3 5.5l-5.492 5.313a2 2 0 0 1-3 .019L5 15c-1.5-1.5-3-3.2-3-5.5"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-2xl font-bold text-red-600"
          data-testid="favorite-count"
        >
          0
        </div>
        <p
          class="text-xs text-muted-foreground"
          data-testid="favorite-percentage"
        >
          0.00
          %
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Strongest
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-zap h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 14a1 1 0 0 1-.78-1.63l9.9-10.2a.5.5 0 0 1 .86.46l-1.92 6.02A1 1 0 0 0 13 10h7a1 1 0 0 1 .78 1.63l-9.9 10.2a.5.5 0 0 1-.86-.46l1.92-6.02A1 1 0 0 0 11 14z"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        >
          Superman
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          Strength:
          10
          /10
        </p>
      </div>
    </div>
    <div
      class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm"
      data-slot="card"
    >
      <div
        class="@container/card-header auto-rows-min grid-rows-[auto_auto] gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 flex flex-row items-center justify-between space-y-0 pb-2"
        data-slot="card-header"
      >
        <div
          class="text-sm font-medium"
          data-slot="card-title"
        >
          Smartest
        </div>
        <svg
          aria-hidden="true"
          class="lucide lucide-trophy h-4 w-4 text-muted-foreground"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M10 14.66v1.626a2 2 0 0 1-.976 1.696A5 5 0 0 0 7 21.978"
          />
          <path
            d="M14 14.66v1.626a2 2 0 0 0 .976 1.696A5 5 0 0 1 17 21.978"
          />
          <path
            d="M18 9h1.5a1 1 0 0 0 0-5H18"
          />
          <path
            d="M4 22h16"
          />
          <path
            d="M6 9a6 6 0 0 0 12 0V3a1 1 0 0 0-1-1H7a1 1 0 0 0-1 1z"
          />
          <path
            d="M6 9H4.5a1 1 0 0 1 0-5H6"
          />
        </svg>
      </div>
      <div
        class="px-6"
        data-slot="card-content"
      >
        <div
          class="text-lg font-bold"
        >
          Batman
        </div>
        <p
          class="text-xs text-muted-foreground"
        >
          Intelligence: 
          10
          /10
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`HeroStats > should render component with default values 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;
